### DISCUSSION TITLE
How can I seed a local file

### DISCUSSION THREAD

**Discussion by bsergean** [anacrolix/torrent#556] [Q&A] [answered]
This is probably simpler as a question than this [convoluted one](https://github.com/anacrolix/torrent/discussions/549).

Inputs:
* I have a local file on disk. 
* I also have a tracker server + url.
* I can run on a machine with a web-server.

The [bittorent spec](http://www.bittorrent.org/beps/bep_0003.html) list those steps.

To start serving, a host goes through the following steps:
1. Start running a tracker (or, more likely, have one running already).
2. Start running an ordinary web server, such as apache, or have one already.
3. Associate the extension .torrent with mimetype application/x-bittorrent on their web server (or have done so already).
4. Generate a metainfo (.torrent) file using the complete file to be served and the URL of the tracker.
5. Put the metainfo file on the web server.
6. Link to the metainfo (.torrent) file from some other web page.
7. Start a downloader which already has the complete file (the 'origin').

Step 4 can be done programmatically.
Step 7 is probably the not so clear step.

2 related questions, with the documentation tag, both a bit abandoned.
* https://github.com/anacrolix/torrent/issues/300
* https://github.com/anacrolix/torrent/issues/306

I think having a solution where one can run a set of commands (from the cmd/) folder would be awesome ; I don't know how people debug but that's what I do.

**Comment by anacrolix** [comment]
Sorry the documentation is very sparse here, it's hard to get time with also being flat out actually implementing stuff 😆 . You can create the torrent file per step 4, by using `./cmd/torrent-create`. You should be able to seed, per step 7, by running `./cmd/torrent download --seed the.step4.torrent`, from the same directory as where step 4 was run (so the files should be in the same place).

**Reply by anacrolix** [reply]
I've uncovered some rough edges in testing this, and am pushing some minor fixes.

**Reply by bsergean** [reply]
Awesome, thanks !

**Reply by anacrolix** [reply]
For testing, you might also find `--testpeer localhost:otherclientport` useful. I've pushed the small fixes I made while testing the steps you described against Transmission to `master`.

**Reply by bsergean** [reply]
Great / I see a v1.31.0 tag too, does that tag contain the fixes ?

**Reply by anacrolix** [reply]
Yes

**Reply by bsergean** [reply]
Awesome, I'll give that a try.

**Reply by bsergean** [reply]
This works great !

I think one problem I had was that I have a company firewall that was preventing the default trackers. Once I ran a local tracker on my mac things started working.

**Reply by anacrolix** [reply]
https://github.com/anacrolix/torrent/issues/248 would be a nice workaround for the issue you had.

**Comment by bsergean** [comment, marked as answer]
# Prerequisite

Make sure you have version [v1.31.0](https://github.com/anacrolix/torrent/releases/tag/v1.31.0) of anacrolix/torrent.

## Step 1 - Compile torrent commands

In a new terminal (1), we will clone the anacrolix repo and generate 2 binaries in /tmp/

```
$ git clone https://github.com/anacrolix/torrent.git
$ cd torrent
torrent$ go build -o /tmp/torrent_download ./cmd/torrent
torrent$ go build -o /tmp/torrent_create ./cmd/torrent-create
```

## Step 2 - Run a local tracker

In a new terminal (2), we will build and run [chihaya](https://github.com/chihaya/chihaya) with the default config.

```
$ git clone git@github.com:chihaya/chihaya.git
$ cd chihaya
$ go build ./cmd/chihaya
$ ./chihaya --config dist/example_config.yaml
INFO[0000] starting metrics server                       addr="0.0.0.0:6880"
INFO[0000] starting storage                              name=memory
INFO[0000] started storage                               gcInterval=3m0s name=memory peerLifetime=31m0s promReportInterval=1s shardCount=1024
INFO[0000] starting tracker logic                        posthooks="[]" prehooks="[]"
INFO[0000] starting HTTP frontend                        addr="0.0.0.0:6969" allowIPSpoofing=false announceRoutes="[/announce]" defaultNumWant=50 enableKeepAlive=false enableRequestTiming=false httpsAddr= idleTimeout=30s maxNumWant=100 maxScrapeInfoHashes=50 readTimeout=5s realIPHeader=x-real-ip scrapeRoutes="[/scrape]" tlsCertPath= tlsKeyPath= writeTimeout=5s
INFO[0000] starting UDP frontend                         addr="0.0.0.0:6969" allowIPSpoofing=false defaultNumWant=50 enableRequestTiming=false maxClockSkew=10s maxNumWant=100 maxScrapeInfoHashes=50 privateKey="paste a random string here that will be used to hmac connection IDs"
...
```

## Step 3 - Generate torrent and seed file with a first client

In a new terminal (3), run the commands below. We will make some temp dirs, and create a new fake file that we will seed.

Notice how we are passing the -n option (to strip default trackers from the generated torrent ; **I need to do that as my mac is behind a firewall**), and then the -a option to add the udp url of the locally running chihaya.

### Generate torrent

```
$ cd /tmp/
/tmp$ mkdir torrent_root
/tmp$ cd torrent_root 
torrent_root$ echo yo > foo.txt
torrent_root$ /tmp/torrent_create -n -a=udp://127.0.0.1:6969 foo.txt > foo.txt.torrent 
```

### Seed file

Run this command **from the same folder**. We pass the --addr :9002 as I get errors with the default port. Your mileage might vary here, change the port if the one below does not work for you.

```
torrent_root$ /tmp/torrent_download download --addr :9002 --seed foo.txt.torrent 
2021-09-02T18:36:01-0700 NONE  main.go:353: downloaded ALL the torrents
1.002704744s: downloading "foo.txt": 3 B/3 B, 1/1 pieces completed (0 partial): 0 B/s
2.003430756s: downloading "foo.txt": 3 B/3 B, 1/1 pieces completed (0 partial): 0 B/s
...
```

## Step 4 - Download the torrent with a second client

In a new terminal (4), navigate to your local anacrolix/torrent repo and download the file.

```
torrent$ go run ./cmd/torrent download --addr :9001 /tmp/torrent_root/foo.txt.torrent 
2021-09-02T19:46:17-0700 NONE  main.go:353: downloaded ALL the torrents
torrent$ cat foo.txt 
yo
```

You downloaded your file !!! 
🍷🍷🍷🍷🍺🍺🍺🍺🍺

**Reply by anacrolix** [reply]
Note a shortcut for step one: `go install github.com/anacrolix/torrent/cmd/...@v1.31.0`. Likely you can do something similar for chihaya.
